### PR TITLE
feat: Tab widget should have navigation arrows when the tabs don't fit

### LINKED ISSUE #3429: [Feature] Tab widget should have navigation arrows when the tabs don't fit
## Summary

When a tab widget has many pages and not enough space to show all of them, there need to be arrows that allow the user to navigate the list and find the tab they need.
A search for tabs could be optional and useful too.

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description
When a tab widget has many pages and not enough space to show all of them, there need to be arrows that allow the user to navigate the list and find the tab they need.
A search for tabs could be optional and useful too.

Fixes #3429 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/3429-tab-add-nav-arrows 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.1 **(0.04)** | 36.58 **(0.02)** | 34.54 **(0.03)** | 55.6 **(0.03)**
 :sparkles: :new: | **app/client/src/widgets/TabsWidget/constants.ts** | **100** | **100** | **100** | **100**
 :red_circle: | app/client/src/widgets/TabsWidget/component/index.tsx | 89.11 **(-10.89)** | 65 **(-18.33)** | 81.82 **(-18.18)** | 90.7 **(-9.3)**
 :green_circle: | app/client/src/widgets/TabsWidget/widget/index.tsx | 78.57 **(1.15)** | 47.14 **(0)** | 78.95 **(0)** | 77.78 **(0.51)**</details>

### CODE REVIEW

**Comment by wmdev0808** [comment]
> @wmdev0808 both arrows are on the same side of the tabs was this the requirement?
> 
> tabs
> 
> @momcilo-appsmith cud you pls do a design review of this scrollable tab header?

@marks0351 , there is no any specific requirements on the design, so it has been implemented by my own decision

File: app/client/src/widgets/TabsWidget/component/index.tsx [hunk 1]
```diff
@@ -116,40 +157,171 @@ const StyledText = styled.div<TabProps>`
...
       tabContainerRef.current?.scrollTo({ top: 0, behavior: "smooth" });
     }
   }, [props.shouldScrollContents]);
 
+  useEffect(() => {
+    setTabsContainerWidth(width);
+  }, [width]);
+
+  useEffect(() => {
+    if (isFirstRender.current) {
```
Comment (marks0351) [inline app/client/src/widgets/TabsWidget/component/index.tsx:218]: why do we need this isFirstRender ref?

File: app/client/src/widgets/TabsWidget/component/index.tsx [hunk 2]
```diff
@@ -116,40 +157,171 @@ const StyledText = styled.div<TabProps>`
...
+  }, [tabs.length]);
+
+  useEffect(() => {
+    if (tabsRef.current) {
+      // Check if the current tabs container has scroll
+      const scrollWidth = tabsRef.current.scrollWidth;
+      const clientWidth = tabsRef.current.clientWidth;
+      if (scrollWidth > clientWidth) {
+        setIsScrollable(true);
+        return;
```
Comment (marks0351) [inline app/client/src/widgets/TabsWidget/component/index.tsx:232]: pls remove return statement and use an if else for readability and consistency.

File: app/client/src/widgets/TabsWidget/component/index.tsx [hunk 3]
```diff
@@ -116,40 +157,171 @@ const StyledText = styled.div<TabProps>`
...
+    const tabsWidth = tabRefs.slice(tabScrollIndex).reduce((total, tabRef) => {
+      const tabWidth = tabRef.current?.scrollWidth || 0;
+      return (total += tabWidth);
+    }, 0);
+    if (tabsWidth + SCROLL_NAV_CONTROL_CONTAINER_WIDTH <= width) {
+      setIsMaxScrolled(true);
+      return;
+    }
+    setIsMaxScrolled(false);
+  }, [tabScrollIndex, tabRefs]);
```
Comment (marks0351) [inline app/client/src/widgets/TabsWidget/component/index.tsx:249]: tabRefs refs don't trigger effect, remove it as dependency and check if the code performs as expected.

**Review by marks0351 — CHANGES_REQUESTED** [review]
resizing widget doesn't seem to properly show scrollable indicators. pls check.
Pls add test cases around this feature.

File: app/client/src/widgets/TabsWidget/component/index.tsx (hunk 1 above)
Comment (wmdev0808) [reply app/client/src/widgets/TabsWidget/component/index.tsx:218]: You're right, that was just dead code, thank you

File: app/client/src/widgets/TabsWidget/component/index.tsx (hunk 3 above)
Comment (wmdev0808) [reply app/client/src/widgets/TabsWidget/component/index.tsx:249]: No, it is required, we are keeping the refs as a state, we  need to calculate visible tabs width based on updated tabRefs state

File: app/client/src/widgets/TabsWidget/component/index.tsx (hunk 1 above)
Comment (jsartisan) [reply app/client/src/widgets/TabsWidget/component/index.tsx:218]: @marks0351 resizing/dragging are very flaky with cypress. In the past, I tried adding a few tests related to dragging/resizing but they were flaky. In the end, had to comment out those tests.

Comment (marks0351) [reply app/client/src/widgets/TabsWidget/component/index.tsx:218]: my bad...i meant to write test cases for the enhancement. we can skip resize cases if it is going to introduce flakiness.

File: app/client/src/widgets/TabsWidget/component/index.tsx (hunk 3 above)
Comment (wmdev0808) [reply app/client/src/widgets/TabsWidget/component/index.tsx:249]: No, it triggers effect.
tabRefs is not a ref, it's just a state(an array of refs), it is required to properly set `isMaxScrolled` state

File: app/client/yarn.lock [hunk 4]
```diff
@@ -3979,18 +3979,18 @@ ansi-escapes@^3.0.0:
   version "3.2.0"
   resolved "https://registry.npmjs.org/ansi-escapes/-/ansi-escapes-3.2.0.tgz"
 
+ansi-escapes@^4.1.0, ansi-escapes@^4.3.0:
```
Comment (marks0351) [inline app/client/yarn.lock:3982]: can u pls check why these yarn dependencies have changed now?

**Review by marks0351 — CHANGES_REQUESTED** [review]
code looks fine to me. 
have a minor comment about the yarn.lock file.
pls move this to QA.

**Comment by somangshu** [comment]
@yatinappsmith @Aishwarya-U-R the Table_Widget_Add_button_spec.js is still failing, Can you guys tell us how to move forward?

**Comment by yatinappsmith** [comment]
@NandanAnantharamu Can you review this failure

**Comment by NandanAnantharamu** [comment]
> @NandanAnantharamu Can you review this failure

@somangshu latest run i see this test is working, i checked failure run looks like intermittent

**Comment by somangshu** [comment]
@NandanAnantharamu can you point where are you seeing the tests passing, I can see the failure in all above actions

**Comment by NandanAnantharamu** [comment]
> @NandanAnantharamu can you point where are you seeing the tests passing, I can see the failure in all above actions

@somangshu https://dashboard.cypress.io/projects/eyxvp8/runs/13594/overview

**Comment by somangshu** [comment]
> https://dashboard.cypress.io/projects/eyxvp8/runs/13594/overview

I see failures here too!

**Comment by NandanAnantharamu** [comment]
> > https://dashboard.cypress.io/projects/eyxvp8/runs/13594/overview
> 
> I see failures here too!

yeah, this is a different failure, looks intermittent . These are not seen in other PR though

**Comment by somangshu** [comment]
RE-running tests in that case

**Review by marks0351 — APPROVED** [review]

**Pull request merged**